bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/api/Authorization.java

```diff
@@ -0,0 +1,533 @@
+            tokenResponseJson.createdOn = Instant.now().toString();
+            // a new refresh token is delivered optional
+            // if not set in response take old one
+            if (Constants.NOT_SET.equals(tokenResponseJson.refreshToken)) {
+                tokenResponseJson.refreshToken = token.getRefreshToken();
+            }
+            token = decodeToken(tokenResponseJson);
+            if (authTokenIsValid()) {
+                String tokenStore = Utils.GSON.toJson(tokenResponseJson);
+                logger.debug("Token result {}", token.toString());
+                storage.put(identifier, tokenStore);
+            } else {
+                token = Utils.INVALID_TOKEN;
+                storage.remove(identifier);
+                logger.warn("Refresh token delivered invalid result {}", tokenResponse);
+            }
+        } else {
+            logger.debug("Token refersh delivered not parsable result {}", tokenResponse);
+            token = Utils.INVALID_TOKEN;
+        }
+    }
+
+    private AccessTokenResponse decodeToken(@Nullable TokenResponse tokenJson) {
+        if (tokenJson != null) {
+            AccessTokenResponse atr = new AccessTokenResponse();
+            atr.setCreatedOn(Instant.parse(tokenJson.createdOn));
+            atr.setExpiresIn(tokenJson.expiresIn);
+            atr.setAccessToken(tokenJson.accessToken);
+            if (!Constants.NOT_SET.equals(tokenJson.refreshToken)) {
+                atr.setRefreshToken(tokenJson.refreshToken);
+            } else {
+                // Preserve refresh token if available
+                if (!Constants.NOT_SET.equals(token.getRefreshToken())) {
+                    atr.setRefreshToken(token.getRefreshToken());
+                } else {
+                    logger.debug("Neither new nor old refresh token available");
+                    return Utils.INVALID_TOKEN;
+                }
+            }
+            atr.setTokenType("Bearer");
+            atr.setScope(Constants.AUTH_SCOPE);
+            return atr;
+        } else {
+            logger.debug("Token Response is null");
+        }
+        return Utils.INVALID_TOKEN;
+    }
+
+    public boolean authTokenIsValid() {
+        return !Constants.NOT_SET.equals(token.getAccessToken()) && !Constants.NOT_SET.equals(token.getRefreshToken());
+    }
+
+    /**
+     * Login to MercedesMe using the resume authentication process. After successful login, the access token is updated.
+     *
+     * @throws MercedesMeAuthException if response status isn't correct or needed parameters not delivered
+     * @throws MercedesMeApiException if an error occurs during API calls
+     */
+    public void login() throws MercedesMeAuthException, MercedesMeApiException {
+        logger.info("Start login");
+        if (isJunit()) {
+            // avoid real API calls in JUnit environment
+            throw new MercedesMeAuthException("Unit Test");
+        }
```

Comment:
Can you extract this test code from the production code? For example by subclassing and overriding the login method.